### PR TITLE
Add a host header to OCSP requests

### PR DESCRIPTION
**Pull request by SnijderC**
Add Host header for OCSP requests, required for some OCSP servers, e.g. Let's Encrypt.

### CODE REVIEW

**Comment by SnijderC** [comment]
To see the issue I'm trying to fix, you can reproduce a request to the Let's Encrypt OCSP server with `openssl`:
``` bash
# Get the certificate chain and extract the issuer and server certificate from the chain
echo "" | \
openssl s_client -connect chrissnijder.nl:443 \
    -servername chrissnijder.nl \
    -showcerts 2>&1  | \
sed  -ne '/-----BEGIN/,/-----END/ p;w chain.pem' | \
sed -ne '1,/-----END/ w server.pem' -e '2,$ !d' -e '/-----BEGIN/,$ w issuer.pem'
# Extract the oscp URL from the server certificate
OCSP_URL=$(openssl x509 -noout -ocsp_uri -in server.pem)
# Send an OCSP request
openssl ocsp -no_nonce -VAfile issuer.pem -issuer issuer.pem -cert server.pem -url $OCSP_URL
```
The above would result in:

``` bash
Error querying OCSP responder
140736391005192:error:27076072:OCSP routines:PARSE_HTTP_LINE1:server response error:ocsp_ht.c:314:Code=400,Reason=Bad Request
```

whereas: 
``` bash
# Extract the hostname from the URL (remove http:// and trailing /)
OCSP_HOST=$(echo $OCSP_URL | sed -e 's/http://' -e 's/\///g')
openssl ocsp -issuer issuer.pem -cert server.pem -url $OCSP_URL -header host $OCSP_HOST
```
results in: 
``` bash
Response verify OK
server.pem: good
	This Update: Feb 26 17:00:00 2017 GMT
	Next Update: Mar  5 17:00:00 2017 GMT
```

**Comment by SnijderC** [comment]
Also at: https://www.openssl.org/docs/man1.0.2/apps/ocsp.html on `-header`

>-header name value
>    If sending a request to an OCSP server, then the specified header name and value are added to the HTTP request. Note that the name and value must be specified as two separate parameters, not as a single quoted string, and that the header name does not have the trailing colon. __*Some OCSP responders require a Host header; use this flag to provide it.*__ [emphasis mine]

**Comment by wbond** [comment]
Thanks, this is a great catch! Would you mind also adding the host header to the CRL client?

**Comment by SnijderC** [comment]
Sure, but...:
 - Where should `_add_header()` and `_get_host()` be placed? Now that 2 modules need the same functions.
 - Why is `crl_client` not already using `_add_header()`?

Duplicating code is not nice, so maybe a new file `_utils.py`?

**Comment by wbond** [comment]
Perhaps `_urllib.py`, containing all of the Python 2/3 abstraction and then using it in the crl and ocsp clients?

**Comment by SnijderC** [comment]
Ha, sorry for the typo in the commit message that made `crl_client()` look like a method in instead of a module. :)

File: certvalidator/ocsp_client.py [hunk 1]
```diff
@@ -9,10 +9,10 @@
 from . import errors
 from ._types import str_cls, type_name
 from .version import __version__
+from ._urllib import _get_host, _add_header
 
 if sys.version_info < (3,):
```
Comment (wbond) [inline certvalidator/ocsp_client.py]: Can we move this `if`/`else` block into `_urllib.py` and import `URLError`, `Request`, `urlopen` from `._urllib` as appropriate in `oscp_client.py` and `crl_client.py`?

I suppose an ideal further change would be to wrap `Request` on Python 2 to automatically call `util.iri_to_uri()`.

With those changes, I believe both clients should be free from know details about what version of Python they are being run on.

**Comment by SnijderC** [comment]
Created a wrapper that makes Request functionally equivalent *for these modules*, I'm not sure this is removing obscurity since we add it right back to another module, but there it is.

**Comment by SnijderC** [comment]
Did you get time to check this out? Would be really nice if you could merge this, then I can remove some workarounds in our project.

**Comment by wbond** [comment]
Sorry, been kind of behind on open source stuff due to other commitments. Hopefully this week.

**Comment by SnijderC** [comment]
Thanks for pulling, I don't mean to push you, but any idea when a new version will be released? If it's going to take some time, I need to implement some workaround to fetch the master branch in our project. Thanks!

**Comment by wbond** [comment]
I was hoping to get #4 merged in before doing a release.

If you are using pip, I think you can point it at a github repo.

**Pull request merged**